### ISSUE TITLE
[1.13] fix Swarm API status codes

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#29408]
Pull request https://github.com/docker/docker/pull/28532 changed the status code for Swarm API endpoints from `406 Not Acceptable` to `503 Service Unavailable`

There are two issues with this change currently, that **must** be addressed / decided on before 1.13 GA

1. **The change in status code must be versioned**

    (see https://github.com/docker/docker/pull/28349#issuecomment-262236124); currently, a docker 1.13 daemon returns status `503 Service Unavailable` for all API versions, including for API 1.24, whereas a docker 1.12 daemon returns status `406`. Given that we cannot change a versioned API, we should update the code to return `406` for API 1.24, and `503` (or what we decide on, see `2.`) for API 1.25 and up

    I don't think there are discussions about this change, it just needs to be implemented.

2. **We must decide on the correct status code**

     See https://github.com/docker/docker/pull/28532#issuecomment-265839863; it is disputed that a `5xx` status code is correct; reasoning there is that if Swarm mode is inactive, the Swarm API endpoints are not implemented, hence "not available". For that reason a `404` status code is suggested. 

    w.r.t. https://github.com/docker/docker/pull/28532#issuecomment-265868008;
    > "It also allows a client to differentiate between missing and not part of the server at all." 

   we should look if there _are_ other actions to be taken by the client; what other actions do we expect the client to take if swarm mode is not enabled; if it's just "print the error message", then this may not be a problem.

ping @stevvooe @justincormack @vieux @bfirsh PTAL. I made this a P0 so that we make the required changes before GA

**Comment by icecrime** [comment]
We already have people looking into what it takes to support services by default, such that you don't ever need to "enable Swarm-mode" to use them.

My 2cts: let's spend our efforts making that happen the sooner rather than debating error codes, and revert back to 404.

Cc @ehazlett.

**Comment by justincormack** [comment]
@icecrime but the only thing we all agree on is that `406` is totally wrong! Literally any other return code is more correct.

**Comment by cpuguy83** [comment]
@justincormack 420?

**Comment by stevvooe** [comment]
I disagree that the status codes need to be versioned. The specification provided no indication of status codes being returned, so any consuming code that makes the assumption of any particular return code would not be correct.

We cannot use 406 here as it is so wrong that it hurts. This was a bug and I am not sure how it ever got in. Doing so would mean that v1.24 violates the HTTP specification.

**Comment by shin-** [comment]
@stevvooe I don't know about the specification, but the [Engine API reference](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.24/#/update-a-swarm) for 1.24 clearly states a status code 406 will be returned if the node is not part of a swarm (or conversely for join / init, if it is already part of a swarm). In my mind, this is pretty clear cut, as this page has been the de facto reference for as long as I can remember when it comes to the API's behavior.

But even if that wasn't compelling enough, you can't really say "takesie-backsies" on something that was published in a GA product. as displeasing as this may be to you or the HTTP purists.

**Comment by stevvooe** [comment]
@shin- I am not sure what you mean, but this isn't about being a purist. This is about avoiding the sloppy engineering process that let this get into the release.

These error codes weren't specified until they were discovered as error codes. Using that retcon as evidence of them being a part of the API is intellectually dishonest. If you want to pretend that consuming these as 406 was part of the intent or design of the API, go right ahead, but any code making that assumption before this _ex post facto_ API addition was broken.

You can't "takesie-backsies" on something that was never given in the first place.

**Comment by shin-** [comment]
But it was given. It's a part of the binary released for general availability.  
Now, you could use the argument that the behavior of the endpoint in that case was unspecified, until you allowed the behavior to be specified in the official documentation (#28349).

It's unfortunate if "sloppy engineering process" led to this, but breaking a versioned API is just as sloppy. In the grand scheme of things, users will care a lot more about their applications not working properly because of an unexpected API change than they will be about an HTTP status code being used incorrectly.

**Comment by stevvooe** [comment]
@shin- I LGTM'd not fully understanding that we were editing an API after the fact. You can see this in both my uneasy response and the fact that __this was in November, three months after a GA release__.

This isn't breaking a versioned API nor is this about using a status code incorrectly; this is clarifying that the behavior of the API was never specified. You wrote what you based on unspecified behavior and now it is broken. In the grand scheme of things, if you make assumptions about an API, and your application breaks, you shouldn't have made assumptions about the API. You can't blame this on the upstream project or a pack of unicorns. You blame that on yourself and move on.

**Comment by shin-** [comment]
I meant you as in "engine maintainers", not you personally.

In the same way, this is not really about my code (we've already made the necessary fixes in docker-py to work around the 1.13-RC changes) but about the expectations of the community and users. If you choose not to take this into account, that's your (again plural) decision. I just want to give you the perception from developers downstream. It doesn't matter **when** this was added to the reference - it's present in the reference today and people have used this information to develop their applications. From this viewpoint, that's all that matters. You have other concerns which I respect and understand, but your ultimate decision should take everything else into account, too.

**Comment by stevvooe** [comment]
@shin- I agree. I don't think this decision took everything into account. However, I disagree that it absolutely matters __when__ a change was made. __We released _completely un-versioned_ change to the API specification, without notice__. That needs to be rolled back.

1. We should have never back-ported status codes to an API specification. If the error codes aren't on the specification, the implementor didn't intend them to be there and we don't know if the behavior was intended or was a bug.
2. It should be clear in the specification that unspecified errors codes may change. We need to be clear about which ones assign general action and which ones have specific action. For example, do you have 4xx or 5xx or specifically 403. The registry specification has this and it has allowed us to migrate status codes where they don't matter. Changes are documented in lettered patch versions.
3. We should have actual error codes, per the `errcode` package the API that can point to specific errors and definitions. Status codes get mapped _by error code_ rather than string parsing.
4. The updated API specifications should have never been released to production. I am not sure how we can make a decision on this if this stuff is just pushed out without thought. Before making a decision about API specifications, we should have a full view of the impact. This is a simple process but some how, API specs just fly out the window.

I'm going to take personal fault for this, as I should have called this out a _lot_ earlier.

**Comment by AkihiroSuda** [comment]
@thaJeztah @stevvooe is this still priority/P0 ?

**Comment by thaJeztah** [comment]
Removed the priority/p0; I added it at the time to have the discussion before we released 1.13
